lib/web/apiserver.go

```diff
@@ -2738,26 +2738,16 @@ func (h *Handler) clusterUnifiedResourcesGet(w http.ResponseWriter, request *htt
 				Logger:            h.log,
 			})
 			unifiedResources = append(unifiedResources, app)
-		case types.AppServerOrSAMLIdPServiceProvider:
-			if r.IsAppServer() {
-				app := ui.MakeApp(r.GetAppServer().GetApp(), ui.MakeAppsConfig{
-					LocalClusterName:  h.auth.clusterName,
-					LocalProxyDNSName: h.proxyDNSName(),
-					AppClusterName:    site.GetName(),
-					Identity:          identity,
-					UserGroupLookup:   getUserGroupLookup(),
-					Logger:            h.log,
-				})
-				unifiedResources = append(unifiedResources, app)
-			} else {
-				app := ui.MakeSAMLApp(r.GetSAMLIdPServiceProvider(), ui.MakeAppsConfig{
-					LocalClusterName:  h.auth.clusterName,
-					LocalProxyDNSName: h.proxyDNSName(),
-					AppClusterName:    site.GetName(),
-					Identity:          identity,
-				})
-				unifiedResources = append(unifiedResources, app)
-			}
+		case types.SAMLIdPServiceProvider:
+			// SAMLIdPServiceProvider resource are shown as
+			// "apps" in the UI.
+			app := ui.MakeAppTypeFromSAMLApp(r, ui.MakeAppsConfig{
+				LocalClusterName:  h.auth.clusterName,
+				LocalProxyDNSName: h.proxyDNSName(),
+				AppClusterName:    site.GetName(),
+				Identity:          identity,
+			})
+			unifiedResources = append(unifiedResources, app)
```

Comment:
I think the same compatibility issue can happen here. If the root cluster is newer than leaf (which would return SAML apps through `types.AppServerOrSAMLIdPServiceProvider`), then the user won't see them.

Reply:
I am failing to see how this affects root and leaf compatibility. This particular change is on proxy. There isn't a new type introduced and all the three types `types.AppServerOrSAMLIdPServiceProvider`, `types.SAMLIdPServiceProvider` and `types.AppServer` are known to auth server for version `15`. 

The resource fetching is on the auth itself: 
This PR: https://github.com/gravitational/teleport/blob/0b2e0d6e6574836688363483af145e197ab080ce/lib/auth/auth_with_roles.go#L1822 and before this PR
existing: https://github.com/gravitational/teleport/blob/6664946bed1d8d5813850d23da414b1a9347be8f/lib/auth/auth_with_roles.go#L4997

The only thing that affect this would be a Connect client that is on upgraded version, but connects to a leaf cluster that is on lower version. From you other comment, looks like I'll have to check for Connect compatibility anyway.

Reply:
Ok, I didn't know how exactly things on the server side work, I thought this change was on auth.